src/corehost/cli/hostpolicy/hostpolicy.cpp

```diff
@@ -22,101 +25,102 @@ namespace
+
+            // Create a CoreCLR instance
+            trace::verbose(_X("CoreCLR path = '%s', CoreCLR dir = '%s'"), context->clr_path.c_str(), context->clr_dir.c_str());
+            auto hr = coreclr_t::create(
+                context->clr_dir,
+                host_path.data(),
+                app_domain_friendly_name,
+                context->coreclr_properties,
+                context->coreclr);
+
+            if (!SUCCEEDED(hr))
+            {
+                trace::error(_X("Failed to create CoreCLR, HRESULT: 0x%X"), hr);
+                rc = StatusCode::CoreClrInitFailure;
+            }
+            else
+            {
+                g_context.reset(context);
+                rc = StatusCode::Success;
+            }
+
+            g_context_initializing.store(false);
         }
 
-        return StatusCode::Success;
+        g_context_cv.notify_all();
+        return rc;
     }
 
-    int create_hostpolicy_context(
+    int get_or_create_hostpolicy_context(
         hostpolicy_init_t &hostpolicy_init,
         const arguments_t &args,
         bool breadcrumbs_enabled,
-        std::shared_ptr<hostpolicy_context_t> &context)
-    {
-
-        std::unique_ptr<hostpolicy_context_t> context_local(new hostpolicy_context_t());
-        int rc = context_local->initialize(hostpolicy_init, args, breadcrumbs_enabled);
-        if (rc != StatusCode::Success)
-            return rc;
-
-        context = std::move(context_local);
-
-        return StatusCode::Success;
-    }
-}
-
-int get_or_create_coreclr(
-    hostpolicy_init_t &hostpolicy_init,
-    const arguments_t &args,
-    host_mode_t mode,
-    std::shared_ptr<coreclr_t> &coreclr)
-{
-    coreclr = g_lib_coreclr.lock();
-    if (coreclr != nullptr)
+        hostpolicy_context_t **context)
     {
-        // [TODO] Validate the current CLR instance is acceptable for this request
+        hostpolicy_context_t *existing_context;
+        {
+            std::unique_lock<std::mutex> lock{ g_context_lock };
+            g_context_cv.wait(lock, [] { return !g_context_initializing.load(); });
 
-        trace::info(_X("Using existing CoreClr instance"));
-        return StatusCode::Success;
-    }
+            existing_context = g_context.get();
+            if (existing_context == nullptr)
+                g_context_initializing.store(true);
+        }
 
-    {
-        std::lock_guard<std::mutex> lock{ g_lib_lock };
-        coreclr = g_lib_coreclr.lock();
-        if (coreclr != nullptr)
+        if (existing_context != nullptr)
         {
-            trace::info(_X("Using existing CoreClr instance"));
-            return StatusCode::Success;
+            // [TODO] Validate the current context is acceptable for this request
+            trace::info(_X("Host context has already been initialized"));
+            assert(existing_context->coreclr != nullptr);
+            *context = existing_context;
```

Comment:
Originally I was planning on creating secondary contexts in hostpolicy as well. I think Steve noted at some point during the discussions around threading that hostpolicy wouldn't really care about secondary contexts, which made sense to me (but I could really go either way).

I guess a benefit of not putting them in hostpolicy is that it doesn't need to do the same kind of tracking/checking as in hostfxr of whether a context is the right type for some operation. And a downside is - as you mentioned - weird handling of properties for secondary contexts.

Reply:
I see - it makes sense. I'm fine to keep it (it makes the hostpolicy API more explicit, which is fine). We just need to describe this somewhere in the code - otherwise it's a bit confusing.

I thought about handling of the properties... and it's OK to have the special case for secondary contexts in my mind. They don't allow setting properties anyway. The only downside is that the property unification algorithm will have to be in hostfxr. So it can't be versioned with the runtime.

I don't know which one would be better long term - one algorithm for all frameworks, or one per runtime. I can see pros for both... but if this would be the only reason to introduce multiple contexts to hostpolicy… it's probably not worth it.

Reply:
> the property unification algorithm will have to be in hostfxr

What is property unification algorithm referring to? Is that the checks on initializing a secondary context for if there are new / different property values in the runtime config compared to the loaded runtime (which I think could/would still happen in hostpolicy)? Or just the tracking of properties for secondary contexts? something else?

Reply:
Yeah - it's my bad name for the "compare properties between the first and secondary context, and report special result if there are differences".
You're right, it could be done in hostpolicy - I missed the fact that corehost_initialize also gets the host interface.

So this means that `corehost_load` is called with the host interface for the first context... and then `corehost_initialize` is called wit the same host interface... right?

Looking at the code it seems that in theory we almost don't need the `g_init` - it's basically equivalent to the `g_context`.
* I find it weird that `corehost_initialize` takes the `init` but for the first context effectively ignores it (since it solely relies on `g_init`).
* `corehost_resolve_component_dependencies` also relies on `g_init`

Maybe it would be too much churn - too risky. So it is probably OK to leave this duplication in the code. But in my mind it is a duplication currently... (but I may be missing something again)

Reply:
Yeah, I'd agree that it is some duplication right now. I think we could/should collapse `g_init` into `g_context`, but that was a bit more than I wanted to do with so little lead/bake time, since it would:
- basically change the first step of every entry to hostpolicy
- add certain use cases for `g_context` that I currently don't have to handle yet (e.g. the `corehost_resolve_component_dependencies` you pointed out, creation/usage without expectation of the runtime ever being loaded, resetting of host command)

Reply:
OK - let's leave it as is for now... I think we should still be able to make the change later on.
It might help to reduce the size of this change...

Reply:
Tried to add comments to make it a bit clearer how `g_init`/`corehost_load` and `g_context`/'corehost_initialize` are currently used.